### PR TITLE
Add a cloud converter object to convert any pre-existing project into a project that will be qfieldcloud compatible

### PR DESCRIPTION
**Pull request by nirvn**
This PR applies QFieldCloud UI/UX related improvements to qfieldsync.

First, the toolbar has been re-ordered and two new QFieldCloud actions have been:
image

**Streamlining cloud project creation from pre-existing projects**

The first action (the white star above the blue cloud) offers users a straightforward way to convert any pre-existing QGIS project to a QFieldCloud-compatible project into a provided destination directory.

The converter does to following:
- converts _all_ vector layers to geopackage (unless the original layer is a geopackage, in which case it simply copies the file) and copies that converted layer into the destination directory
- copies all file-based layers that aren't vectors into the destination directory

Much like the offline converter (i.e.the preexisting cable export), the converter takes care of updating data source paths. The conversion to geopackage will respect joins, virtual fields, and subset/SQL filters.

After the file conversion/copying is done, the converted cloud project is automatically uploaded to the user's QFieldCloud account

The converter dialog opens to this:
image

When the user hits the [ create ] button (I've had a bit of a dilemma settling on convert vs. create, I welcome your thoughts there), and the following progress bars are keeping track of remaining tasks:

image

The dialog will warn users of typical pitfalls to avoid (insuring a unique cloud project name, insuring a clean directory, etc.),

When the user is over with this conversion, he/she can grab his/her mobile device, log into QFieldCloud and grab the project.

**Improvement to currently open cloud project synchronizing**

The second action added to the toolbar is a synchronization button that enables itself when the currently opened project in QGIS is detected as being the local copy of a cloud project. The button leads directly to the synchronization dialog.

The dialog itself has received a fair amount of improvement. First, instead of doing communication to<->from the server in the background, the dialog is opened straight away and tells to user to grab some pop corn while project file details are being fetched:
image

The dialog then moves onto the list of files in need of being synchronized. That part hasn't been changed from what we had before:
image

I have a couple of ideas on how to improve things there, but this PR is already too big so I'll defer to a different PR once this is done.

When the user has settled on what to synchronize, he/she hits the [ sync! ] button, and is shown the progress panel. Some improvements there include only showing the upload/download progress bar if there is actually some uploads and/or downloads taking place. Some wording cleanups were applied too (avoiding stuff like file(s), intimidating words like 'overwriting', etc.:
image

Finally, when the synchronization is over, the dialog ends with:
image

If a user tries to synchronize a local cloud project that's already fully sync'ed, this is what he/she is presented with:
image

There are a bunch of little tweaks and improvements through the code too. For e.g., when the currently open cloud project is synchronized, we automatically reload the project. Some quality of life improvements all around

**Preferences panel: QFieldCloud has arrived**

Lastly, this PR re-organizes the plugin's preferences panel. A screenshot should be self-explanatory:
image

### CODE REVIEW

**Comment by nirvn** [comment]
This feature requires libqfieldsync to include the following PR (https://github.com/opengisch/libqfieldsync/pull/5).

File: qfieldsync/core/cloud_converter.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                if not layer_source.is_supported:
+                    project.removeMapLayer(layer)
+                    continue
+
+                if layer.dataProvider() is not None:
+                    md = QgsProviderRegistry.instance().providerMetadata(
+                        layer.dataProvider().name()
+                    )
+                    if md is not None:
+                        decoded = md.decodeUri(layer.source())
```
Comment (suricactus) [inline qfieldsync/core/cloud_converter.py:96]:
```suggestion
                    provider_meta = QgsProviderRegistry.instance().providerMetadata(
                        layer.dataProvider().name()
                    )
                    if provider_meta is not None:
                        decoded = provider_meta.decodeUri(layer.source())
```

File: qfieldsync/core/cloud_converter.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                        if "path" in decoded:
+                            path = pathResolver.writePath(decoded["path"])
+                            if path.startswith("localized:"):
+                                # layer stored in localized data path, skip
+                                continue
+
+                if layer.type() == QgsMapLayer.VectorLayer:
+                    if not layer_source.convert_to_gpkg(self.export_folder):
```
Comment (suricactus) [inline qfieldsync/core/cloud_converter.py:104]: So there is NO way to keep you pg layer a PG layer, even if it is a public postgis?

Btw, if it fails, would be nice to have some kind of UI indication

**Comment by nirvn** [comment]
@suricactus , just pushed the UI part here. Please savagely review the code, I'll address while you guys are sleeping :wink:

File: qfieldsync/core/cloud_api.py [hunk 3]
```diff
@@ -610,6 +610,20 @@ def projects(self) -> Optional[List[CloudProject]]:
...
+    def is_currently_open_project_cloud_local(self) -> bool:
+        project_dir = QgsProject.instance().homePath()
+
+        found = False
+        for project_id, local_dir in self.preferences.value(
+            "qfieldCloudProjectLocalDirs"
+        ).items():
+            if local_dir == project_dir:
+                found = True
+                break
```
Comment (m-kuhn) [inline qfieldsync/core/cloud_api.py:623]:
```suggestion
        local_project_dirs = self.preferences.value("qfieldCloudProjectLocalDirs").items()
        found = any((project_dir == local_dir for _, local_dir in local_project_dirs))
```

Some python yoga?
Your call which version is more readable

File: qfieldsync/gui/cloud_converter_dialog.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        self.qfield_preferences = Preferences()
+        self.network_manager = network_manager
+
+        if not self.network_manager.has_token():
+            login_dlg = CloudLoginDialog(self.network_manager, self)
+            login_dlg.authenticate()
+            login_dlg.accepted.connect(
+                lambda: self.network_manager.projects_cache.refresh()
+            )
+            login_dlg.rejected.connect(lambda: self.close())
```
Comment (m-kuhn) [inline qfieldsync/gui/cloud_converter_dialog.py:65]: Can we deduplicate this code?

File: qfieldsync/gui/cloud_converter_dialog.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        else:
+            self.network_manager.projects_cache.refresh()
+
+        project_name = self.project.baseName()
+        if project_name:
+            pattern = re.compile("[\W_]+")  # NOQA
```
Comment (m-kuhn) [inline qfieldsync/gui/cloud_converter_dialog.py:71]:
```suggestion
            pattern = re.compile(r"[\W_]+")
```

Or is this really meant to be this way?

File: qfieldsync/gui/cloud_converter_dialog.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            self.project, self.get_export_folder_from_dialog()
+        )
+
+        cloud_convertor.warning.connect(self.show_warning)
+        cloud_convertor.total_progress_updated.connect(self.update_total)
+
+        try:
+            cloud_convertor.convert()
+        except Exception:
+            QApplication.restoreOverrideCursor()
```
Comment (m-kuhn) [inline qfieldsync/gui/cloud_converter_dialog.py:145]: https://github.com/opengisch/QgisModelBaker/blob/master/QgisModelBaker/utils/qt_utils.py#L240-L249

Super nice utility class there. Except async, then it becomes more complicated :laughing: 

```py
with OverrideCursor(Qt.WaitCursor):
    do_something_and_restore_cursor_no_matter_what()
```

File: qfieldsync/ui/preferences_widget.ui [hunk 7]
```diff
@@ -15,50 +15,76 @@
...
      </property>
-     <layout class="QGridLayout" name="generalGridLayout">
-      <item row="0" column="0">
-       <widget class="QLabel" name="importDirectoryLabel">
+     <layout class="QGridLayout" name="cloudGridLayout">
+      <item row="0" column="0" colspan="2">
+       <widget class="QLabel" name="cloudIntroLabel">
         <property name="text">
-         <string>Default import directory</string>
+         <string>&lt;span style=&quot; font-style:italic;&quot;&gt;&lt;a href=&quot;https://qfield.cloud/&quot;&gt;QFieldCloud&lt;/a&gt; allows your team to focus on what's important, making sure you efficiently get quality field data. Thanks to its integration with QField, you and your team will be able to start surveying and digitising data in no time.&lt;/span&gt;</string>
```
Comment (m-kuhn) [inline qfieldsync/ui/preferences_widget.ui:26]:
```suggestion
         <string>&lt;span style=&quot; font-style:italic;&quot;&gt;&lt;a href=&quot;https://qfield.cloud/&quot;&gt;QFieldCloud&lt;/a&gt; allows your team to focus on what's important, making sure you efficiently get quality field data. Thanks to its integration with QField, you and your team will be able to start surveying and digitising data collaboratively in no time.&lt;/span&gt;</string>
```

File: qfieldsync/ui/cloud_transfer_dialog.ui [hunk 8]
```diff
@@ -19,6 +19,65 @@
...
+       <property name="rightMargin">
+        <number>0</number>
+       </property>
+       <property name="bottomMargin">
+        <number>0</number>
+       </property>
+       <item>
+        <widget class="QLabel" name="projectFilesLabel">
+         <property name="text">
+          <string>Please wait while project files details are fetched from QFieldCloud…</string>
```
Comment (m-kuhn) [inline qfieldsync/ui/cloud_transfer_dialog.ui:39]:
```suggestion
          <string>Please wait while project file details are fetched from QFieldCloud…</string>
```

Or is the plural justified?

**Review by m-kuhn — COMMENTED** [review]
Need to test it !

**Review by suricactus — APPROVED** [review]
:rocket: awesome!

**Pull request merged**